### PR TITLE
Include status 30X for URL unshortening

### LINKED ISSUE #1114: Link unshortener fails to shorten several URLs
- Issue type: Bug report

### Short description

Loklak fails to unshorten many URLs, most of them are of type `https://fb.me/*`.

### Environment

 - Operating system: CentOS 7
 - Software version: Even with `loklak:development`.
 - Software source: Github Repository

### Steps to reproduce

1. Run loklak
```bash
$ ./bin/start.sh
```
2. Look at the log
```bash
$ tail -f data/loklak.log | less
```
3. You can see that unshorten fails for many URLs.

### Expected behaviour
The links should be shortened.

### Actual behaviour
Shortening links fail.

### PR DESCRIPTION
**Pull request by singhpratyush**
### Include HTTP status `30X` while unshortening
Since any of the HTTP `30X` is may be used by many URL shortening services, it should be also considered while unshortening.

Fixes #1114.

I have:
- [x] There is a corresponding issue for this pull request.
- [x] Mentioned the Issue number in the pull request commit message `Fixes #<number> commit message`
- [x] There is only strictly only one commit per issue.

### For the reviewers
I have:
- [ ] Reviewed this pull request by an authorized contributor.
- [ ] The reviewer is assigned to the pull request.

### CODE REVIEW

**Comment by mariobehling** [comment]
@singhpratyush Thanks for your PR. Where has this been discussed? I don't see an issue. Please open an issue. We really want to get you on board, but it is difficult to work together if you don't follow best practices, especially as we have not worked together a lot. 

Could you please help reviewers and match issues and PRs? I could write a long comment here now, but I do hope you understand yourself how important it is, that we make it easy for reviewers and collaborators to follow guidelines.

Please - do - read FOSSASIA's best practices, which I have send to you before: http://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia/#commits

Please help us and follow best practices.

Best practices are the basis of our work in upcoming months.

It is absolutely vital, that you follow best practices.

Thank you.

**Comment by mariobehling** [comment]
Just to be sure: There are a number of different people that will work on loklak projects over upcoming months. Please: Do follow Best Practices.

Maybe, it is not so clear: We need to follow Best Practices together.

Thank you and let's follow best practices together. The first step: Please read them. Second step: Let's adhere to Best Practices together.

Everyone appreciates that you join development and make it easy for all of us to collaborate following -> Best Practices.

Thanks! Let's do this (following Best Practices). Cheers.

**Comment by singhpratyush** [comment]
@mariobehling: I missed the issue number in PR description but it is there in the commit message. Sorry about that. I'll make sure that I don't miss out the next time.

**Comment by yukiisbored** [comment]
@singhpratyush Please rebase your branch. I recommend to do it manually from git than using the magic green button that Github provides.

**Comment by singhpratyush** [comment]
@yukiisbored: I have rebased and build is successful.

**Comment by yukiisbored** [comment]
@singhpratyush Please rebase again. Sorry :/

File: src/org/loklak/http/ClientConnection.java [hunk 1]
```diff
@@ -272,7 +273,7 @@ public static String getRedirect(String urlstring, boolean useAuthentication) th
                 throw new IOException("redirect for  " + urlstring+ ": no location attribute found");
             } else {
                 EntityUtils.consumeQuietly(httpEntity);
-                throw new IOException("no redirect for  " + urlstring+ " fail: " + httpResponse.getStatusLine().getStatusCode() + ": " + httpResponse.getStatusLine().getReasonPhrase());
+                return urlstring;
```
Comment (sudheesh001) [inline src/org/loklak/http/ClientConnection.java]: This is to understand redirects, Can you explain why it doesn't throw an `IOException` anymore and instead can you ensure it always returns the `urlstring`?

Comment (singhpratyush) [inline src/org/loklak/http/ClientConnection.java]: [This](https://github.com/loklak/loklak_server/blob/2abbf30e1a96f278ae06472d4f3ed98151c237ff/src/org/loklak/harvester/RedirectUnshortener.java#L76) is the function that calls `ClientConnection#getRedirect()` to get shortened URL. The `IOException` (which I removed) would be raised only if the status is non `30X`, which will be caught in caller function and the [latest rendered URL will be returned](https://github.com/loklak/loklak_server/blob/2abbf30e1a96f278ae06472d4f3ed98151c237ff/src/org/loklak/harvester/RedirectUnshortener.java#L89).

Now, if we replace the exception throwing by simply returning the `urlstring` (which is not semantically an issue as we tried to check redirect of a URL which doesn't redirect anywhere), the caller function will [break](https://github.com/loklak/loklak_server/blob/2abbf30e1a96f278ae06472d4f3ed98151c237ff/src/org/loklak/harvester/RedirectUnshortener.java#L82) and our job is done.

When the exception is thrown, the caller function panics due to it and [raises a false alarm in the logs](https://github.com/loklak/loklak_server/blob/2abbf30e1a96f278ae06472d4f3ed98151c237ff/src/org/loklak/harvester/RedirectUnshortener.java#L88). That's why I removed it.

But as I mentioned in #1114, there are cases when the response is non `30X` but redirect is made using `meta` tag. This can be tackled by extracting URL from the tag. But I think that it should go in another patch, hence haven't added it.

Comment (sudheesh001) [inline src/org/loklak/http/ClientConnection.java]: Can you add it into this itself, lets tackle all these at once.

Comment (singhpratyush) [inline src/org/loklak/http/ClientConnection.java]: Ok. Shall I amend it?

**Comment by singhpratyush** [comment]
@sudheesh001: While making the changes you suggested, I came across a situation where `https?://fb.me/*` links fail to unshorten, giving `400` error. While playing around a bit, I found that this was due to the `GET` request that was being made. While the `POST` requests did things well (I tested this on `requests` and `httpie` too).

So I introduced two separate methods to make `GET` and `POST` requests. If `GET` request would fail with non `200` error, an attempt will be made to unshorten using `POST` request.

Upon making this change, all the URLs are getting unshortened (from what I have observed till now).

Also, making only `POST` requests is not a good idea. Because some services (Tumblr for instance) consider it as a non-genuine request (give `403`). So, `POST` shall be used only if `GET` fails.

**Comment by mariobehling** [comment]
@singhpratyush I am not sure, if your additional changes should be discussed in an issue first. This might help to speed up the review process. Fix one thing at a time.

**Comment by singhpratyush** [comment]
@mariobehling: I agree that there are many changes introduced in this PR but they all fix the same thing.

@Orbiter @sudheesh001: Shall I break it into 3 parts?
1. Accepting all `30X` codes.
2. Fetching redirect URL from `meta` tag.
3. Falling back to `POST` request in case of errors.

**Comment by singhpratyush** [comment]
@sudheesh001: Sounds good. I will make the changes.

**Comment by singhpratyush** [comment]
@sudheesh001: I have made the changes. I was missing the base class for `HttpPost` and `HttpGet` earlier and hence decided to put them in different functions altogether. But such implementation was already there in the same file and it helped.

**Comment by singhpratyush** [comment]
Please review.

File: src/org/loklak/http/ClientConnection.java [hunk 2]
```diff
@@ -241,44 +244,127 @@ private void init() throws IOException {
...
+        if (indexContent < 0)
+            return null;
+        html = html.substring(indexContent);
+        int indexURLStart = html.indexOf("http");
+        if (indexURLStart < 0)
+            return null;
+        html = html.substring(indexURLStart);
+        int indexURLEnd = html.indexOf("\"");
+        if (indexURLEnd < 0)
+            return null;
```
Comment (vibhcool) [inline src/org/loklak/http/ClientConnection.java]: how about using regex here?

File: src/org/loklak/http/ClientConnection.java [hunk 3]
```diff
@@ -241,44 +244,127 @@ private void init() throws IOException {
...
+                return getRedirect(urlstring, useAuthentication, false);
             } else {
+                if (debugLog)
+                    Log.getLog().info("Trying to fetch Meta redirect URL for " + urlstring);
+                String metaURL = getMetaRedirectURL(httpEntity);
                 EntityUtils.consumeQuietly(httpEntity);
-                throw new IOException("no redirect for  " + urlstring+ " fail: " + httpResponse.getStatusLine().getStatusCode() + ": " + httpResponse.getStatusLine().getReasonPhrase());
+                if (metaURL != null) {  // A URL was found in meta tag
+                    if (debugLog)
+                        Log.getLog().info("Fetched Meta redirect URL for " + urlstring + " : " + metaURL);
```
Comment (vibhcool) [inline src/org/loklak/http/ClientConnection.java]: you can replace this by DAO.log

File: src/org/loklak/http/ClientConnection.java [hunk 4]
```diff
@@ -241,44 +244,127 @@ private void init() throws IOException {
...
         if (httpEntity != null) {
-            if (httpResponse.getStatusLine().getStatusCode() == 301) {
-                for (Header header: httpResponse.getAllHeaders()) {
-                    if (header.getName().equalsIgnoreCase("location")) {
-                        EntityUtils.consumeQuietly(httpEntity);
-                        return header.getValue();
-                    }
-                }
+            int httpStatusCode = httpResponse.getStatusLine().getStatusCode();
+            if (300 <= httpStatusCode && httpStatusCode <= 308) {
```
Comment (vibhcool) [inline src/org/loklak/http/ClientConnection.java]: if httpStatusCode <=300
then it will be less than 308

File: src/org/loklak/http/ClientConnection.java (hunk 2 above)
Comment (singhpratyush) [inline src/org/loklak/http/ClientConnection.java]: It will give the correct results but will be slower. This method is not hard to follow so there will be no problem in maintaining it.

File: src/org/loklak/http/ClientConnection.java (hunk 4 above)
Comment (singhpratyush) [inline src/org/loklak/http/ClientConnection.java]: `httpStatusCode` is not `<= 300`.

File: src/org/loklak/http/ClientConnection.java (hunk 3 above)
Comment (singhpratyush) [inline src/org/loklak/http/ClientConnection.java]: I've changed it.

**Comment by singhpratyush** [comment]
I have made some more improvements to the patch - 
* Improved meta redirect URL extractor.
* Modified unshortening method to shorten links until a good spot (no `30X`, no meta redirect) is reached, instead of stopping at any which is not applicable to shortening.

I merged this and #1173 branches to [this](https://github.com/singhpratyush/loklak_server/tree/temp_unshorten) branch and [tests are passing](https://travis-ci.org/singhpratyush/loklak_server/builds/234151221).

**Comment by mariobehling** [comment]
@singhpratyush How can we speed up the process on this PR? Expected flow is that a PR should not take more than a day or should be split up in smaller issues/PRs. Let's follow this on the next PR. Please ask your co-developers to review to get this rolling. Thanks.

**Comment by singhpratyush** [comment]
@vibhcool @SarishiNoHara @kavithaenair @djmgit @SKrPl: Can you guys please build this locally and test to verify proper functioning?

**Comment by singhpratyush** [comment]
@mariobehling: I agree that this PR should have been broken down, and will do so for future patchs.

As of now, I have tested it against the tests introduces in #1173 and they passed.

**Comment by Orbiter** [comment]
I have not tested the code but it looks smart and should be good to merge if peer review is ok.
